shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -84,22 +90,46 @@ public DartMessengerTaskQueue makeBackgroundTaskQueue() {
     }
   }
 
+  /** A serial task queue that can run on a concurrent ExecutorService. */
   private static class DefaultTaskQueue implements DartMessengerTaskQueue {
     @NonNull private final ExecutorService executor;
+    @NonNull private final ConcurrentLinkedQueue<Runnable> queue;
+    @NonNull private final AtomicBoolean isRunning;
 
-    DefaultTaskQueue() {
-      // TODO(gaaclarke): Use a shared thread pool with serial queues instead of
-      // making a thread for each TaskQueue.
-      ThreadFactory threadFactory =
-          (Runnable runnable) -> {
-            return new Thread(runnable, "DartMessenger.DefaultTaskQueue");
-          };
-      this.executor = Executors.newSingleThreadExecutor(threadFactory);
+    DefaultTaskQueue(ExecutorService executor) {
+      this.executor = executor;
+      queue = new ConcurrentLinkedQueue<>();
+      isRunning = new AtomicBoolean(false);
     }
 
     @Override
     public void dispatch(@NonNull Runnable runnable) {
-      executor.execute(runnable);
+      queue.add(runnable);
+      executor.execute(
```

Comment:
There's no way to know if this will be the same thread every time, which means plugins may get `onMethodCall` happening on any number of threads over the course of their lifetime.

Reply:
That's correct.  We specifically stated that in the documentation.  This will only be a problem for `ThreadLocal` usage with is rare.  We can expand makeBackgroundTaskQueue to have a variant that takes in options that can specify if we want a dedicated thread.  This is the default settings where just serial execution will cover 99% of use cases.

Reply:
If you want I could do that in this PR,  allow you to specify if you want: [concurrent, serial, single-thread].  I don't think there is going to anybody that really needs single-thread though.  It will be a bit of a pain for iOS too since that isn't an out of the box option.

Reply:
I remember this being a particular concern for @stuartmorgan - perhaps I'm misunderstanding though.

It's not just an issue for thread local storage - it means that you could, in theory, get concurrent invocations of `onMethodCall` for the same plugin, and then end up with races where two threads try to read/write memory that isn't thread safe.

Reply:
I mostly convinced myself that it should be fine to allow arbitrary non-stable threads as long as the calls are managed by a serial task queue; that means that plugin code doesn't have to be thread safe, it just can't be thread hostile, which is a much lower bar. (Assuming that all its method channels share the same serial queue, but IIRC the API allows for that already).

I have one lingering concern, which is async calls: I think it may not be entirely obvious that if a plugin developer calls an async function—say, one provided by the OS which calls a callback on the calling thread—that the callback code could run in parallel with the other plugin code. We'll need to document that very prominently, and keep an eye out for whether it ends up being a problem in practice.

Reply:
> it means that you could, in theory, get concurrent invocations of onMethodCall for the same plugin

The serialization of the tasks happening here should make that impossible.  Let me know if you think there is a path through the code that violates that.

> (Assuming that all its method channels share the same serial queue, but IIRC the API allows for that already)

Yep, that's an option available.  You can create a TaskQueue and send it into multiple channels.

> I have one lingering concern, which is async calls:

Yep, that could be a problem.  It is a problem that already exists with regular platform channels too though fwiw.  It's standard threading gotcha fare.

Let me know if you guys want to discuss further over VC.

Reply:
> > I have one lingering concern, which is async calls:
> 
> Yep, that could be a problem. It is a problem that already exists with regular platform channels too though fwiw. It's standard threading gotcha fare.

No, the problem I'm talking about doesn't exist with regular platform channels. While it's absolutely possible to create threading complexity for yourself deliberately with existing platform channels, I'm talking about a more subtle case that is broken by a thread pool. Take the following imaginary iOS API (I know this is Android, but I'm more familiar with async patterns in ObjC so it'll be easier for me to explain correctly):

```
// Does foo, then calls completion with the result. The completion handler will be called on the
// same thread as the initial call.
- (void)doFooWithCompletion:(void (^)(SomeResultObject result));
```

I can then write the following plugin code right now without introducing any threading considerations in my code:
```
...
  [thing doFooWithCompletion:^(SomeResultObject result)(SomeResultObject result) {
    self.lastFooResult = result;
    ...
  }];
```

That same code suddenly becomes unsafe if I use background method channels, and the only solution (if we don't provide a stable-thread option) is to make all of the relevant code threadsafe.

Reply:
> The completion handler will be called on the same thread as the initial call.

Ahh this is the part I was missing.  I don't know that I've see an API with that convention.  Usually they are either [callback on main handler, callback on some background handler, callback on provided handler].  In order to "call completion lambda on same handler that issued the message" you'd have to have a way to access the handler that issued the message which typically isn't available unless you pass it in.  (See also discussions about deprecating `dispatch_get_current_queue` https://stackoverflow.com/a/17678556).

Reply:
It may be that since it's now been quite some time since Apple deprecated getting the current queue it's no longer a thing people do on iOS, in which case the wait-and-see approach to figuring out if we need to provide an easy solution should work out fine.

Reply:
Ok.

In that case, it sounds like we might be concerned if someone was able to dispatch on the TaskQueue itself and deadlock themselves, but it sounds like we're not making dispatch public so it's probably ok.

Reply:
@dnfield I want to open that up so people can run initialization code on the task queue when plugins are setup.  What's your concern there?  I was going to do that in a separate PR but I might do it here to make testing easier.

Reply:
They will block themselves on the completion of a dispatched task and it will not get scheduled because they're currently blocking the queue, but the method itself makes that very not-obvious.

This is not a huge problem in small, simple plugins maintained by one developer. It will be a bigger problem in larger multi-team maintained plugins.

Reply:
Okay, I'll postpone that for now and keep it simple.